cpp/src/arrow/compute/kernels/vector_selection_test.cc

```diff
@@ -2365,38 +2365,91 @@ TEST_F(TestDropNullKernelWithTable, DropNullTableWithSlices) {
   });
 }
 
-TEST(TestIndicesNonZero, IndicesNonZero) {
+template <typename Type>
+class TestIndicesNonZero : public ::testing::Test {};
+
+TYPED_TEST_SUITE(TestIndicesNonZero, NumericArrowTypes);
+TYPED_TEST(TestIndicesNonZero, IndicesNonZero) {
   Datum actual;
   std::shared_ptr<Array> result;
+  auto type = TypeTraits<TypeParam>::type_singleton();
 
-  ASSERT_OK_AND_ASSIGN(
-      actual,
-      CallFunction("indices_nonzero", {ArrayFromJSON(uint32(), "[null, 50, 0, 10]")}));
+  ASSERT_OK_AND_ASSIGN(actual, CallFunction("indices_nonzero",
+                                            {ArrayFromJSON(type, "[null, 50, 0, 10]")}));
   result = actual.make_array();
-  AssertArraysEqual(*result, *ArrayFromJSON(uint64(), "[1, 3]"));
+  AssertArraysEqual(*ArrayFromJSON(uint64(), "[1, 3]"), *result);
+
+  // empty
+  ASSERT_OK_AND_ASSIGN(actual,
+                       CallFunction("indices_nonzero", {ArrayFromJSON(type, "[]")}));
+  result = actual.make_array();
+  AssertArraysEqual(*ArrayFromJSON(uint64(), "[]"), *result);
+
+  // chunked
+  ChunkedArray chunked_arr(
+      {ArrayFromJSON(type, "[1, 0, 3]"), ArrayFromJSON(type, "[4, 0, 6]")});
+  ASSERT_OK_AND_ASSIGN(
+      actual, CallFunction("indices_nonzero", {static_cast<Datum>(chunked_arr)}));
+  AssertArraysEqual(*ArrayFromJSON(uint64(), "[0, 2, 3, 5]"), *actual.make_array());
+
+  // empty chunked
+  ChunkedArray chunked_arr_empty({ArrayFromJSON(type, "[1, 0, 3]"),
+                                  ArrayFromJSON(type, "[]"),
+                                  ArrayFromJSON(type, "[4, 0, 6]")});
+  ASSERT_OK_AND_ASSIGN(
+      actual, CallFunction("indices_nonzero", {static_cast<Datum>(chunked_arr_empty)}));
+  AssertArraysEqual(*ArrayFromJSON(uint64(), "[0, 2, 3, 5]"), *actual.make_array());
+}
 
+TEST(TestIndicesNonZero, IndicesNonZeroBoolean) {
+  Datum actual;
+  std::shared_ptr<Array> result;
+
+  // boool
   ASSERT_OK_AND_ASSIGN(
       actual, CallFunction("indices_nonzero",
                            {ArrayFromJSON(boolean(), "[null, true, false, true]")}));
   result = actual.make_array();
   AssertArraysEqual(*result, *ArrayFromJSON(uint64(), "[1, 3]"));
+}
 
-  ASSERT_OK_AND_ASSIGN(actual,
-                       CallFunction("indices_nonzero",
-                                    {ArrayFromJSON(float64(), "[null, 1.3, 0.0, 5.0]")}));
+TEST(TestIndicesNonZero, IndicesNonZeroDecimal) {
+  Datum actual;
+  std::shared_ptr<Array> result;
+
+  ASSERT_OK_AND_ASSIGN(
+      actual, CallFunction(
+                  "indices_nonzero",
+                  {DecimalArrayFromJSON(decimal128(2, -2), R"(["12E2",null,"0","0"])")}));
   result = actual.make_array();
-  AssertArraysEqual(*result, *ArrayFromJSON(uint64(), "[1, 3]"));
+  AssertArraysEqual(*ArrayFromJSON(uint64(), "[0]"), *result, true);
```

Comment:
Please make sure that arguments are intelligible:

```suggestion
  AssertArraysEqual(*ArrayFromJSON(uint64(), "[0]"), *result, /*verbose=*/true);
```

Reply:
done